<a href="https://colab.research.google.com/github/Vishvesh-Bhardwaj/Machine-Learning-Core-Concepts/blob/main/RA2011026010109_MLCC_PCA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**RA2011026010109 - VISHVESH BHARDWAJ - Principal component analysis**

Let's use the famous Iris dataset, which contains measurements of sepal length, sepal width, petal length, and petal width for **150 iris flowers, with 50 observations* for each of the three species:** ***Iris setosa, Iris versicolor, and Iris virginica.**

**First, let's load the dataset and split it into training and testing sets:**

In [1]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=42)

**Next, let's create a logistic regression model and train it on the original dataset:**

In [2]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()
lr.fit(X_train, y_train)

print("Accuracy on training set:", lr.score(X_train, y_train))
print("Accuracy on testing set:", lr.score(X_test, y_test))

Accuracy on training set: 0.975
Accuracy on testing set: 1.0


/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


This gives us an accuracy of around **97.5%** on the training set and **100%** on the testing set, which indicates that the ***model is overfitting.***

Now, **let's apply PCA to the dataset to reduce its dimensionality.**
We'll keep only the first two principal components, which should capture most of the variance in the data:

In [3]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

**Finally, let's train the logistic regression model on the reduced dataset:**

In [4]:
lr_pca = LogisticRegression()
lr_pca.fit(X_train_pca, y_train)

print("Accuracy on training set (after PCA):", lr_pca.score(X_train_pca, y_train))
print("Accuracy on testing set (after PCA):", lr_pca.score(X_test_pca, y_test))


Accuracy on training set (after PCA): 0.9583333333333334
Accuracy on testing set (after PCA): 1.0


This gives us an accuracy of around **95% on the training set and 100% on the testing set**, which indicates that the model is still performing well but with ***less overfitting.***

Overall, **we can see that applying PCA to the Iris dataset led to a reduction in the dimensionality of the data while maintaining high accuracy on the testing set.**

*However, the reduction in dimensionality also led to a slight decrease in accuracy on the training set, which is to be expected when we discard some of the information in the original dataset.*